packages/tailwindcss/src/index.ts

```diff
@@ -482,6 +485,10 @@ export async function compile(
           return compiledCss
         }
 
+        // Arbitrary values (`text-[theme(--color-red-500)]`) and arbitrary
+        // properties (`[--my-var:theme(--color-red-500)]`) can contain function
+        // calls so we need evaluate any functions we find there that weren't in
+        // the source CSS.
```

Comment:
see comment above but this is where `theme()` calls output by utilities (doing like addUtilities or matchUtilities) are processed.